### PR TITLE
change to use BinaryAllocator to replace BinaryAllocatorFunctor

### PR DESCRIPTION
**Pull request by johnfangAFW**
Change to use a BinaryAllocator to replace the original BinaryAllocatorFunctor due to the following reasons:
1) holds a default block size, which the function could not have such a state
2) provide a method to allocate by a specific size instead of the default block size

All unit tests and integration tests passed.

root@b0d49a77c3ce:/build# cd build/test && ctest
Test project /build/build/test
    Start 1: plogmock
1/9 Test #1: plogmock .........................   Passed    0.38 sec
    Start 2: persistent_volume
2/9 Test #2: persistent_volume ................   Passed    3.09 sec
    Start 3: transport
3/9 Test #3: transport ........................   Passed    0.09 sec
    Start 4: readcache
4/9 Test #4: readcache ........................   Passed    0.03 sec
    Start 5: skv_record
5/9 Test #5: skv_record .......................   Passed    0.05 sec
    Start 6: key_encoding
6/9 Test #6: key_encoding .....................   Passed    0.04 sec
    Start 7: skv_ser
7/9 Test #7: skv_ser ..........................   Passed    0.04 sec
    Start 8: expression
8/9 Test #8: expression .......................   Passed    0.05 sec
    Start 9: transport
9/9 Test #9: transport ........................   Passed    0.04 sec

100% tests passed, 0 tests failed out of 9

Total Test time (real) =   3.96 sec

### CODE REVIEW

**Review by iafuture — CHANGES_REQUESTED** [review]
The proposed solution here does solve the problem but I think we should keep the lambda-based approach as it is more efficient and quite a bit more flexible. 

The efficiency issue is mostly due to the now required allocations from the shared_ptr usage. You can avoid that by accepting a BinaryAllocator instance instead.

The bigger issue though is that this approach is not as flexible. The user of Payload now only gets to specify what size blocks are allocated but not how or where they come from. With a lambda, I can have you allocate memory from a designated area, such as SCM, or DPDK-marked memory, or even from a different thread. To achieve the same with the class approach, you'll have to now build inheritance-based solution which gets things even more inefficient due to vtables.

I'd revert back to the lambda-based approach with a default size in the TCP and RDMA implementations(VirtualNetworkStack.cpp), as we discussed in the online chat. Let's do another chat if you have more questions

**Comment by johnfangAFW** [comment]
> The proposed solution here does solve the problem but I think we should keep the lambda-based approach as it is more efficient and quite a bit more flexible.
> 
> The efficiency issue is mostly due to the now required allocations from the shared_ptr usage. You can avoid that by accepting a BinaryAllocator instance instead.
> 
> The bigger issue though is that this approach is not as flexible. The user of Payload now only gets to specify what size blocks are allocated but not how or where they come from. With a lambda, I can have you allocate memory from a designated area, such as SCM, or DPDK-marked memory, or even from a different thread. To achieve the same with the class approach, you'll have to now build inheritance-based solution which gets things even more inefficient due to vtables.
> 
> I'd revert back to the lambda-based approach with a default size in the TCP and RDMA implementations(VirtualNetworkStack.cpp), as we discussed in the online chat. Let's do another chat if you have more questions

As discussed, changed to pass both default size and the std::function to BinaryAllocator to avoid inheritance.

File: src/k2/transport/Payload.cpp [hunk 1]
```diff
@@ -36,7 +36,7 @@ Payload::PayloadPosition::PayloadPosition(_Size bIdx, _Size bOff, size_t offset)
     // Empty payload will have position of (bidx=0, boff=0, off=0)
 }
 
-Payload::Payload(BinaryAllocatorFunctor allocator):
+Payload::Payload(std::shared_ptr<BinaryAllocator> allocator):
```
Comment (jfunston) [inline src/k2/transport/Payload.cpp:39]: Can we pass the allocator by move instead of a shared pointer?

**Review by jfunston — CHANGES_REQUESTED** [review]
Maybe you can also run the test_k23si_tpcc.sh test locally and see if there is any noticeable performance degradation, even though I don't expect there to be.

File: src/k2/transport/Payload.cpp (hunk 1 above)
Comment (johnfangAFW) [reply src/k2/transport/Payload.cpp:39]: I chose to use a shared pointer here to give users the flexibility to either use an individual allocator for a payload or use a shared allocator for multiple payloads. If we have the use case to use the same allocator for many payloads, the shared pointer is more efficient so that we don't have to create an allocator instance for each payload.

Run test_k23si_tpcc.sh locally and compare it with the master, didn't see any performance degradation. The times are almost the same.

Comment (jfunston) [reply src/k2/transport/Payload.cpp:39]: I see. If we want to keep the shared pointer we can change it to a seastar::lw_shared_ptr which is faster. The restrictions on it are that there is no polymorphism and it stays on the same core, which we should meet for this use case.

Comment (johnfangAFW) [reply src/k2/transport/Payload.cpp:39]: I see, will change to use seastar::lw_shared_ptr then.

Comment (johnfangAFW) [reply src/k2/transport/Payload.cpp]: done

File: src/k2/transport/Payload.cpp [hunk 2]
```diff
@@ -36,7 +36,7 @@ Payload::PayloadPosition::PayloadPosition(_Size bIdx, _Size bOff, size_t offset)
     // Empty payload will have position of (bidx=0, boff=0, off=0)
 }
 
-Payload::Payload(BinaryAllocatorFunctor allocator):
+Payload::Payload(seastar::lw_shared_ptr<BinaryAllocator> allocator):
```
Comment (iafuture) [inline src/k2/transport/Payload.cpp:39]: I think you should use allocator objects instead of shared pointer. The shared_ptr approach is more expensive since it requires allocation

**Review by iafuture — CHANGES_REQUESTED** [review]

File: src/k2/transport/Payload.cpp (hunk 2 above)
Comment (johnfangAFW) [reply src/k2/transport/Payload.cpp:39]: The purpose of a shared pointer here is to allow multiple payloads to use the same allocator to avoid allocating a BinaryAllocator for each individual payload. It is more expensive to use an allocator object for each payload in such a use case.

Comment (iafuture) [reply src/k2/transport/Payload.cpp:39]: I think the shared_ptr is more expensive. A shared pointer copies a pointer and increments a shared counter. The functor object on the other hand is just a copy of a pointer (the function underneath).

Moreover, the shared pointer does require allocation, as you do in the rest of the code base (calling make_lw_shared). You could "share" the shared_ptr, but this is just as expensive since just copying the object since it is so light.
Lastly, the shared_ptr approach also now requires 2x dereferencing (1 for the shared_ptr, 2 for the function in the object) whereas with the object it is just 1.

Comment (johnfangAFW) [reply src/k2/transport/Payload.cpp:39]: I see, this is because the BinaryAllocator object is so light that the direct object copy here is less expensive than copying a shared pointer and updating the object reference count in the shared pointer.

Will change to use the object directly.

Comment (johnfangAFW) [reply src/k2/transport/Payload.cpp]: done

File: src/k2/transport/VirtualNetworkStack.cpp [hunk 3]
```diff
@@ -24,6 +24,7 @@ Copyright(c) 2020 Futurewei Cloud
 #include "VirtualNetworkStack.h"
 
 // third-party
+#include <memory>
```
Comment (iafuture) [inline src/k2/transport/VirtualNetworkStack.cpp:27]: probably don't need this

File: src/k2/transport/TXEndpoint.cpp [hunk 4]
```diff
@@ -72,18 +72,18 @@ std::unique_ptr<TXEndpoint> TXEndpoint::fromURL(const String& url, BinaryAllocat
...
     K2LOG_D(log::tx, "dtor");
 }
 
-TXEndpoint::TXEndpoint(String&& pprotocol, String&& pip, uint32_t pport, BinaryAllocatorFunctor&& allocator):
+TXEndpoint::TXEndpoint(String&& pprotocol, String&& pip, uint32_t pport, BinaryAllocator allocator):
     protocol(std::move(pprotocol)),
     ip(std::move(pip)),
     port(pport),
-    _allocator(std::move(allocator)) {
+    _allocator(allocator) {
```
Comment (iafuture) [inline src/k2/transport/TXEndpoint.cpp:86]: you should still move here

File: src/k2/transport/TXEndpoint.cpp [hunk 5]
```diff
@@ -72,18 +72,18 @@ std::unique_ptr<TXEndpoint> TXEndpoint::fromURL(const String& url, BinaryAllocat
...
-    return std::make_unique<TXEndpoint>(std::move(protocol), std::move(ip), port, std::move(allocator));
+    return std::make_unique<TXEndpoint>(std::move(protocol), std::move(ip), port, allocator);
 }
 
 TXEndpoint::~TXEndpoint() {
     K2LOG_D(log::tx, "dtor");
 }
 
-TXEndpoint::TXEndpoint(String&& pprotocol, String&& pip, uint32_t pport, BinaryAllocatorFunctor&& allocator):
+TXEndpoint::TXEndpoint(String&& pprotocol, String&& pip, uint32_t pport, BinaryAllocator allocator):
```
Comment (iafuture) [inline src/k2/transport/TXEndpoint.cpp:82]: receive the allocator by && to avoid a copy

File: src/k2/transport/TXEndpoint.h [hunk 6]
```diff
@@ -44,7 +44,7 @@ class TXEndpoint {
 public: // lifecycle
     // construct an endpoint from a url with the given allocator
     // Returns nullptr if there was a problem parsing the url
-    static std::unique_ptr<TXEndpoint> fromURL(const String& url, BinaryAllocatorFunctor&& allocator);
+    static std::unique_ptr<TXEndpoint> fromURL(const String& url, BinaryAllocator allocator);
```
Comment (iafuture) [inline src/k2/transport/TXEndpoint.h:47]: why this change? you should still receive by && to avoid a copy

File: src/k2/transport/TXEndpoint.cpp [hunk 7]
```diff
@@ -38,7 +38,7 @@ namespace k2 {
 // must have: protocol(group1), ip(group2==ipv4, group3==ipv6), port(group4)
 const std::regex urlregex{"(.+)://(?:([^:\\[\\]]+)|\\[(.+)\\]):(\\d+)"};
 
-std::unique_ptr<TXEndpoint> TXEndpoint::fromURL(const String& url, BinaryAllocatorFunctor&& allocator) {
+std::unique_ptr<TXEndpoint> TXEndpoint::fromURL(const String& url, BinaryAllocator allocator) {
```
Comment (iafuture) [inline src/k2/transport/TXEndpoint.cpp:41]: receive by && to avoid copy

File: src/k2/transport/Payload.h [hunk 8]
```diff
@@ -165,7 +165,10 @@ class Payload {
     Payload shareRegion(size_t startOffset, size_t nbytes);
 
     // Creates a new payload as a copy of this payload. The underlying data is copied over to the new payload
-    Payload copy(BinaryAllocatorFunctor dest_allocator=DefaultAllocator);
+    Payload copy(BinaryAllocator dest_allocator);
```
Comment (iafuture) [inline src/k2/transport/Payload.h:168]: since we're keeping this allocator, i think you should receive by &&

File: src/k2/transport/Payload.cpp [hunk 9]
```diff
@@ -434,8 +434,9 @@ Payload Payload::shareRegion(size_t startOffset, size_t nbytes){
     return shared;
 }
 
-Payload Payload::copy(BinaryAllocatorFunctor allocator) {
-    Payload copied(allocator);
+Payload Payload::copy(BinaryAllocator allocator) {
```
Comment (iafuture) [inline src/k2/transport/Payload.cpp:437]: receive by && and then move later on into copied._allocator

**Review by iafuture — CHANGES_REQUESTED** [review]

File: src/k2/transport/VirtualNetworkStack.cpp (hunk 3 above)
Comment (johnfangAFW) [reply src/k2/transport/VirtualNetworkStack.cpp]: removed

File: src/k2/transport/TXEndpoint.cpp (hunk 5 above)
Comment (johnfangAFW) [reply src/k2/transport/TXEndpoint.cpp]: done. But we need to manually copy the allocator in some object for a shared allocator to avoid making the original allocator as empty.

File: src/k2/transport/TXEndpoint.h (hunk 6 above)
Comment (johnfangAFW) [reply src/k2/transport/TXEndpoint.h]: done

File: src/k2/transport/Payload.h (hunk 8 above)
Comment (johnfangAFW) [reply src/k2/transport/Payload.h]: done

File: src/k2/transport/Payload.cpp (hunk 9 above)
Comment (johnfangAFW) [reply src/k2/transport/Payload.cpp]: done, but we need to make a copy of a shared allocator in some cases to avoid making the original allocator empty/invalid.

File: src/k2/transport/TXEndpoint.cpp (hunk 7 above)
Comment (johnfangAFW) [reply src/k2/transport/TXEndpoint.cpp]: done

File: src/k2/transport/TXEndpoint.cpp (hunk 4 above)
Comment (johnfangAFW) [reply src/k2/transport/TXEndpoint.cpp]: done

**Review by jfunston — DISMISSED** [review]

File: src/k2/transport/Payload.cpp [hunk 10]
```diff
@@ -409,7 +409,8 @@ Payload Payload::shareRegion(size_t startOffset, size_t nbytes){
     seek(startOffset);
     nbytes = std::min(_size - _currentPosition.offset, nbytes);
 
-    Payload shared(_allocator);
+    BinaryAllocator copied_allocator = _allocator;
+    Payload shared(std::move(copied_allocator));
```
Comment (iafuture) [inline src/k2/transport/Payload.cpp:413]: I think this could just be
```suggestion
    Payload shared(BinaryAllocator(_allocator));
```

Comment (johnfangAFW) [reply src/k2/transport/Payload.cpp]: done, thanks for suggestions.

**Review by iafuture — APPROVED** [review]
